File: Implement_strStr().cpp
#include <cstring>
using std::strlen;

class Solution {
public:
    const char *strStr(const char *haystack, const char *needle) {
        int pos = sunday(haystack, needle);
        return (pos == -1) ? nullptr : haystack + pos;
    }

private:
    int sunday(const char *src, const char *dst) {
        int slen = strlen(src);
        int dlen = strlen(dst);

        int next[256];
        for (int i = 0; i < 256; ++i) {
            next[i] = dlen + 1;
        }
        for (int i = 0; i < dlen; ++i) {
            next[dst[i]] = dlen - i;
        }

        int pos = 0;
        while (pos < (slen - dlen) + 1) {
            bool found = true;
            for (int i = pos, j = 0; j < dlen; ++j, ++i) {
                if (src[i] != dst[j]) {
                    pos += next[src[pos + dlen]];
                    found = false;
                    break;
                }
            }
            if (found) {
                return pos;
            }
        }
        return -1;
    }
};

File: Implement_strStr().py
class Solution:

    def strStr(self, haystack, needle):
        '''
        Find a substring (needle) in haystack.

        :param haystack: haystack
        :type haystack: str
        :param needle: needle
        :type needle: str
        :return: a string or None

        Usage:
            >>> s = Solution()
            >>> s.strStr("example hello", "example")

        '''
        pos = self.sunday(haystack, needle)
        return haystack[pos:] if pos != -1 else None

    def sunday(self, src, dst):
        next = {c: len(dst) - idx for idx, c in enumerate(dst)}
        pos = 0
        while pos < len(src) - len(dst) + 1:
            for idx, c in enumerate(dst):
                if c != src[pos + idx]:
                    if pos + len(dst) >= len(src):
                        return -1
                    jump_c = src[pos + len(dst)]
                    pos += len(dst) + 1 if jump_c not in next else next[jump_c]
                    break
            else:
                return pos
        return -1
